### PR TITLE
(CONT-366) Syntax update

### PR DESCRIPTION
**Pull request by LukasAud**
Code now following the rules regarding:

legacy facts
top-scope facts
top-scope variables
relative classname inclusion
parameter types
parameter documentation
selector inside resource

### CODE REVIEW

**Comment by LukasAud** [comment]
Current failures are present in the nightly run as well.

**Comment by LukasAud** [comment]
All current test failures are also present in the nightly checks. PR does not seem to implement any new failure.

File: manifests/config/context/resource.pp [hunk 1]
```diff
@@ -18,15 +18,15 @@
...
 define tomcat::config::context::resource (
-  Enum['present','absent'] $ensure = 'present',
-  $resource_name                   = $name,
-  $resource_type                   = undef,
-  $catalina_base                   = $::tomcat::catalina_home,
-  Hash $additional_attributes      = {},
-  Array $attributes_to_remove      = [],
-  Boolean $show_diff               = true,
+  Enum['present','absent'] $ensure    = 'present',
+  String $resource_name               = $name,
```
Comment (smortex) [inline manifests/config/context/resource.pp:22]:
```suggestion
  String[1] $resource_name            = $name,
```

File: manifests/config/context/resource.pp [hunk 2]
```diff
@@ -18,15 +18,15 @@
...
+  Optional[String[1]] $resource_type  = undef,
+  String $catalina_base               = $tomcat::catalina_home,
```
Comment (smortex) [inline manifests/config/context/resource.pp:24]:
```suggestion
  Stdlib::Absolutepath $catalina_base = $tomcat::catalina_home,
```

File: manifests/config/context/resourcelink.pp [hunk 3]
```diff
@@ -19,14 +19,14 @@
...
 define tomcat::config::context::resourcelink (
   Enum['present','absent'] $ensure = 'present',
-  $catalina_base                   = $::tomcat::catalina_home,
-  $resourcelink_name               = $name,
-  $resourcelink_type               = undef,
-  Hash $additional_attributes      = {},
-  Array $attributes_to_remove      = [],
-  Boolean $show_diff               = true,
+  String $catalina_base                   = $tomcat::catalina_home,
+  String $resourcelink_name               = $name,
```
Comment (smortex) [inline manifests/config/context/resourcelink.pp:23]:
```suggestion
  Stdlib::Absolutepath $catalina_base     = $tomcat::catalina_home,
  String[1] $resourcelink_name            = $name,
```

File: manifests/config/context/resources.pp [hunk 4]
```diff
@@ -13,16 +13,16 @@
 #
 define tomcat::config::context::resources (
   Enum['present','absent'] $ensure = 'present',
-  $catalina_base                   = $::tomcat::catalina_home,
+  String $catalina_base            = $tomcat::catalina_home,
```
Comment (smortex) [inline manifests/config/context/resources.pp:16]:
```suggestion
  Stdlib::Absolutepath $catalina_base = $tomcat::catalina_home,
```

File: manifests/config/context/valve.pp [hunk 5]
```diff
@@ -27,16 +27,16 @@
...
-  $class_name                      = undef,
-  $catalina_base                   = $::tomcat::catalina_home,
-  Hash $additional_attributes      = {},
-  Array $attributes_to_remove      = [],
-  Array $uniqueness_attributes     = [],
-  Boolean $show_diff               = true,
+  Optional[String[1]] $resource_name  = undef,
+  Optional[String[1]] $resource_type  = undef,
+  Optional[String[1]] $class_name     = undef,
+  String $catalina_base               = $tomcat::catalina_home,
```
Comment (smortex) [inline manifests/config/context/valve.pp:33]:
```suggestion
  Stdlib::Absolutepath $catalina_base = $tomcat::catalina_home,
```

File: manifests/config/properties/property.pp [hunk 6]
```diff
@@ -8,9 +8,9 @@
 #   The name of the property. `$name`.
 #
 define tomcat::config::properties::property (
-  $catalina_base,
-  $value,
-  $property = $name,
+  String    $catalina_base,
```
Comment (smortex) [inline manifests/config/properties/property.pp:11]:
```suggestion
  Stdlib::Absolutepath $catalina_base,
```

File: manifests/config/properties/property.pp [hunk 7]
```diff
@@ -8,9 +8,9 @@
...
+  String[1] $value,
+  String    $property = $name,
```
Comment (smortex) [inline manifests/config/properties/property.pp:13]:
```suggestion
  String[1] $property = $name,
```

File: manifests/config/server/connector.pp [hunk 8]
```diff
@@ -24,22 +24,22 @@
...
 define tomcat::config::server::connector (
-  $catalina_base                             = undef,
+  Optional[String[1]] $catalina_base         = undef,
   Enum['present','absent'] $connector_ensure = 'present',
-  $port                                      = undef,
-  $protocol                                  = $name,
-  $parent_service                            = 'Catalina',
+  Optional[String[1]] $port                  = undef,
+  String $protocol                           = $name,
+  String $parent_service                     = 'Catalina',
```
Comment (smortex) [inline manifests/config/server/connector.pp:31]:
```suggestion
  String[1] $protocol                        = $name,
  String[1] $parent_service                  = 'Catalina',
```

File: manifests/config/server/engine.pp [hunk 9]
```diff
@@ -30,26 +31,26 @@
...
-  Enum['present','absent'] $class_name_ensure                 = 'present',
-  $engine_name                                                = undef,
-  $jvm_route                                                  = undef,
-  Enum['present','absent'] $jvm_route_ensure                  = 'present',
-  $parent_service                                             = 'Catalina',
-  $start_stop_threads                                         = undef,
-  Enum['present','absent'] $start_stop_threads_ensure         = 'present',
-  $server_config                                              = undef,
-  Boolean $show_diff                                          = true,
+  String $default_host,
```
Comment (smortex) [inline manifests/config/server/engine.pp:34]:
```suggestion
  String[1] $default_host,
```

File: manifests/config/server/engine.pp [hunk 10]
```diff
@@ -30,26 +31,26 @@
...
+  Optional[String[1]] $catalina_base                                = undef,
+  Optional[Variant[Integer, String[1]]] $background_processor_delay = undef,
+  Enum['present','absent'] $background_processor_delay_ensure       = 'present',
+  Optional[String[1]] $class_name                                   = undef,
+  Enum['present','absent'] $class_name_ensure                       = 'present',
+  Optional[String[1]] $engine_name                                  = undef,
+  Optional[String[1]] $jvm_route                                    = undef,
+  Enum['present','absent'] $jvm_route_ensure                        = 'present',
+  String $parent_service                                            = 'Catalina',
```
Comment (smortex) [inline manifests/config/server/engine.pp:43]:
```suggestion
  String[1] $parent_service                                         = 'Catalina',
```

File: manifests/config/server/globalnamingresource.pp [hunk 11]
```diff
@@ -18,16 +18,16 @@
...
 define tomcat::config::server::globalnamingresource (
-  $catalina_base                   = $::tomcat::catalina_home,
-  $resource_name                   = undef,
-  $type                            = 'Resource',
-  Enum['present','absent'] $ensure = 'present',
-  Hash $additional_attributes      = {},
-  Array $attributes_to_remove      = [],
-  $server_config                   = undef,
-  Boolean $show_diff               = true,
+  String $catalina_base               = $tomcat::catalina_home,
```
Comment (smortex) [inline manifests/config/server/globalnamingresource.pp:21]:
```suggestion
  Stdlib::Absolutepath $catalina_base = $tomcat::catalina_home,
```

File: manifests/config/server/globalnamingresource.pp [hunk 12]
```diff
@@ -18,16 +18,16 @@
...
+  Optional[String[1]] $resource_name  = undef,
+  String $type                        = 'Resource',
```
Comment (smortex) [inline manifests/config/server/globalnamingresource.pp:23]:
```suggestion
  String[1] $type                     = 'Resource',
```

File: manifests/config/server/listener.pp [hunk 13]
```diff
@@ -22,18 +22,18 @@
 #   Specifies display differences when augeas changes files, defaulting to true. Valid options: true or false.
 #
 define tomcat::config::server::listener (
-  $catalina_base                            = $::tomcat::catalina_home,
+  String $catalina_base                     = $tomcat::catalina_home,
```
Comment (smortex) [inline manifests/config/server/listener.pp:25]:
```suggestion
  Stdlib::Absolutepath $catalina_base = $tomcat::catalina_home,
```

File: manifests/config/context/manager.pp [hunk 14]
```diff
@@ -17,13 +17,13 @@
 #
 define tomcat::config::context::manager (
   Enum['present','absent'] $ensure = 'present',
-  $catalina_base                   = $::tomcat::catalina_home,
-  $manager_classname               = $name,
+  String $catalina_base            = $tomcat::catalina_home,
+  String $manager_classname        = $name,
```
Comment (smortex) [inline manifests/config/context/manager.pp:21]:
```suggestion
  Stdlib::Absolutepath $catalina_base = $tomcat::catalina_home,
  String[1] $manager_classname     = $name,
```

File: manifests/config/context/parameter.pp [hunk 15]
```diff
@@ -15,14 +15,14 @@
...
 define tomcat::config::context::parameter (
-  Optional[String]                            $value          = undef,
-  Enum['present', 'absent']                   $ensure         = 'present',
-  Pattern[/^(\/[^\/ ]*)+\/?$/]                $catalina_base  = $::tomcat::catalina_home,
-  String                                      $parameter_name = $name,
-  Optional[String]                            $description    = undef,
-  Optional[Boolean]                           $override       = undef,
+  Optional[String[1]]          $value          = undef,
+  Enum['present', 'absent']    $ensure         = 'present',
+  Pattern[/^(\/[^\/ ]*)+\/?$/] $catalina_base  = $tomcat::catalina_home,
```
Comment (smortex) [inline manifests/config/context/parameter.pp:20]:
```suggestion
  Stdlib::Absolutepath $catalina_base          = $tomcat::catalina_home,
```

File: manifests/config/context/parameter.pp [hunk 16]
```diff
@@ -15,14 +15,14 @@
...
+  String                       $parameter_name = $name,
```
Comment (smortex) [inline manifests/config/context/parameter.pp:21]:
```suggestion
  String[1]                    $parameter_name = $name,
```

File: manifests/config/server/tomcat_users.pp [hunk 17]
```diff
@@ -24,25 +24,25 @@
...
-  $element_name                    = undef,
-  Enum['present','absent'] $ensure = present,
-  $file                            = undef,
-  Boolean $manage_file             = true,
-  $owner                           = undef,
-  $group                           = undef,
-  Optional[Variant[String, Sensitive[String]]] $password = undef,
-  Array $roles                     = [],
-  Boolean $show_diff               = true,
+  String $catalina_base                                   = $tomcat::catalina_home,
```
Comment (smortex) [inline manifests/config/server/tomcat_users.pp:27]:
```suggestion
  Stdlib::Absolutepath $catalina_base                     = $tomcat::catalina_home,
```

File: manifests/config/context/environment.pp [hunk 18]
```diff
@@ -25,17 +25,17 @@
 #
 define tomcat::config::context::environment (
   Enum['present','absent'] $ensure    = 'present',
-  Stdlib::Absolutepath $catalina_base = $::tomcat::catalina_home,
+  Stdlib::Absolutepath $catalina_base = $tomcat::catalina_home,
   String $environment_name            = $name,
```
Comment (smortex) [inline manifests/config/context/environment.pp:29]:
```suggestion
  String[1] $environment_name         = $name,
```

File: manifests/config/server/realm.pp [hunk 19]
```diff
@@ -28,25 +28,25 @@
...
-  $parent_engine                                          = 'Catalina',
-  $parent_host                                            = undef,
-  $parent_realm                                           = undef,
-  Hash $additional_attributes                             = {},
-  Array $attributes_to_remove                             = [],
-  Optional[Boolean] $purge_realms                         = undef,
-  $server_config                                          = undef,
-  Boolean $show_diff                                      = true,
+  Optional[String[1]] $catalina_base      = undef,
+  String $class_name                      = $name,
```
Comment (smortex) [inline manifests/config/server/realm.pp:32]:
```suggestion
  String[1] $class_name                   = $name,
```

**Review by smortex — COMMENTED** [review]
2 minor issues I found:
* in defined classes, values that defaults to `$name` should probably be non-empty string (`String[1]`) and not just `String`;
* values that default to `catalina_home` path should probably always be `Stdlib::Absolutepath` (currently a mix of `Stdlib::Absolutepath`, `String` and `Pattern`).

**Comment by LukasAud** [comment]
@smortex Thanks for reviewing. I will investigate and address the $name parameter to make sure its datatype is more accurate. As for the $catalina_home and $catalina_base parameters, they seem to be a bit more delicate. Last time I tried to standardise them, it resulted in a massive flood of failures. It may take a bit more time but I will try again to get them to work.

File: examples/catalina_attributes.pp [hunk 20]
```diff
@@ -1,6 +1,8 @@
+# @param catalina_base
+# @param show_diff
 define tomcat::config::example (
-  $catalina_base     = '/tmp',
-  Boolean $show_diff = true,
+  String[1] $catalina_base = '/tmp',
```
Comment (smortex) [inline examples/catalina_attributes.pp:4]: This can also be a `Stdlib::Absolutepath`

**Review by smortex — COMMENTED** [review]
That looks great :heart:  !

File: examples/catalina_attributes.pp (hunk 20 above)
Comment (LukasAud) [reply examples/catalina_attributes.pp]: 👍

**Review by smortex — APPROVED** [review]
LGTM!

**Review by david22swan — APPROVED** [review]
LGTM

**Pull request merged**